Add type hinting for device and driver modules

WhatTheFuzz opened the pull request:
I'm using a downstream application of this library and got curious how the strings and bytes were handled differently in the Device class. I added some type hints for my own edification that adheres to `mypy --strict`. I also added a couple test cases to ensure it is behaving as intended when using bytes.

WhatTheFuzz committed:
Add type hinting for device.py.

WhatTheFuzz committed:
Add additional tests for writing bytes.

Need clarifaction if this is the intended behavior, as readlines
states: "This [method] only works for mode='t' on Python3".

WhatTheFuzz committed:
Add type hints to driver.

WhatTheFuzz committed:
Resolve reimport.

Driver does not import the requested FtdiError. Rather, it comes
from _base. Import it directly instead of trying to reimport.

codedstructure commented:
Thanks for this PR - I'll take a closer look over the next couple of days.

I do have to say I'm not a big fan of type hinting in Python (I would replace 'best' in your take [here](https://github.com/WhatTheFuzz#spicy-topics) with a different word 😉), but I do see some value and I'm not against this PR on that basis. I do wonder if aiming to add it only where it helps readability & understanding would be better though - e.g. I really can't see how adding `-> None` to `__del__` methods is an improvement - is that something required with the `--strict` mode of mypy?

WhatTheFuzz commented:
Totally a fair take and I'm happy to amend to the project needs. The primary benefit to implementing type checking on methods that return `None` (such as constructors, or as your mentioned, `__del__`) is that it informs the type checker to specifically examine those functions. Without typing the return (and without adding `--strict`), the type checker ignores the function entirely. This could be problematic if the `__del__` method returns an object unexpectedly and you didn't have a test case that handles that. Instead, we're explicitly saying that `__del__` deletes the item and does nothing more. I imagine there is an opposite use case where a maintainer would instead want to return the item that was deleted. Or potentially a boolean.

codedstructure commented on pylibftdi/driver.py:
```diff
@@ -72,36 +73,43 @@ class Driver(object):
         """
         :param libftdi_search: force a particular version of libftdi to be used
             can specify either library name(s) or path(s)
         :type libftdi_search: string or sequence of strings
```
Key question here is whether this alternate is worth preserving - you've effectively made the 'or sequence of strings' no longer valid (previously `_load_library` would take None, str, or List[str] - but AFAICT your intent is to have it only take `None` or `str`). The use-case for multiple strings is to allow a user to have multiple options (e.g. to support multiple OS platforms) but still have some restrictions (e.g. only listing possible library names for libftdi 0.x). The use case for a single string is when the user knows exactly the path / library name they are using (maybe they've built it themselves), and the use-case for None is that it 'just works' in many situations.

Previously the values of `self._lib_search` could be either str or List[str] and these were just transparently passed to `_load_library`, but I think for type consistency in the `_load_library` interface it would be good at this point to still accept `Optional[str]|List[str]` and if it's a `str` then wrap it in a singleton list, so `self._lib_search` is `Optional[List[str]]`

WhatTheFuzz replied:
Got it, I understand the use cases now. I think I was thrown off by "sequence" and wondered how that would be represented by downstream applications. I haven't addressed `_load_library` yet, but I'll fix that typing as well.

codedstructure commented on pylibftdi/driver.py:
```diff
@@ -72,36 +73,43 @@ class Driver(object):
 
         :param name: library name
-        :param search_list: sequence or string referring to library names
+        :param search_list: string referring to library names
```
as in previous comments, I think having Optional[List[str]] at this point satisfies more use-cases. I don't particularly mind at this point in the code, since `_load_library` is an internal function, but I think that's probably the cleanest approach.

WhatTheFuzz replied:
Agreed. I've made the appropriate changes to make the type explicitly an `Optional[list[str]]`.

codedstructure commented on pylibftdi/driver.py:
```diff
@@ -72,36 +73,43 @@ class Driver(object):
         """
+        # If no search list is given, use the default.
+        if not isinstance(search_list, str):
+            raise TypeError(f'library search list must be a string, not {type(search_list)}')
```
this is hit if `None` is passed in, confirmed by running `python3 -m pylibftdi.examples.info` - doesn't need a physical FTDI device attached, just libftdi installed. I think this line could / should be removed and just assume search_list is either `None` or `List[str]` (see earlier comments) and either assigned to `libraries` on line 109 (without wrapping in a list) or the search lookup done (line 107)

WhatTheFuzz replied:
Thanks for catching that. I've added a test in `tests/test_driver.py` that should catch that in the future. The latest commit checks that `search` list is None or a list, otherwise raises a TypeError. Checked that `python3 -m pylibftdi.examples.info` returns the appropriate information.

codedstructure commented on pylibftdi/driver.py:
```diff
@@ -72,36 +73,43 @@ class Driver(object):
-            search_list = self._lib_search.get(name, ())
-        if isinstance(search_list, (str, bytes)):
-            search_list = (search_list,)
+            libraries = self._lib_search.get(name, "")
```
default here should be `[]` for type consistency

WhatTheFuzz replied:
Fixed.

codedstructure commented on pylibftdi/driver.py:
```diff
@@ -114,12 +122,11 @@ def _load_library(self, name, search_list=None):
         if lib is None:
-            raise LibraryMissingError('{} library not found (search: {})'.format(
-                name, search_list))
+            raise LibraryMissingError(f'{name} library not found (search: {str(search_list)})')
```
previous to introducing the `libraries` variable, this string included `search_list` which was the thing being iterated over - should probably still do that (maybe by reverting the introduction of `libraries` and just replacing your lines 104-109 with:

```python
if search_list is None:
    search_list = self._lib_search.get(name, [])
```

and that would be it...

WhatTheFuzz replied:
I implemented your suggestion (really a reversion to what you previously had anyway. :grimacing:)

codedstructure commented on pylibftdi/device.py:
```diff
@@ -7,6 +7,7 @@
 pylibftdi: https://github.com/codedstructure/pylibftdi
 
 """
+from __future__ import annotations
```
This requires Python3.7+, whereas the setup.py classifiers claims support for Python3.6. I should probably remove the more specific classifiers anyway, and since Python3.6 is EOL not too worried by this, doesn't need changing unless you want to remove the specific Python3.6 entry in setup.py

WhatTheFuzz replied:
Bumped the versions specified in the classifiers to support 3.7 to 3.11. Added the `python_requires=">=3.7"` line to be explicit.

codedstructure commented on pylibftdi/device.py:
```diff
@@ -543,24 +546,28 @@ def readline(self, size=0):
             next_char = self.read(1)
             if next_char == '' or (0 < size < len(line_buffer)):
                 break
+            if isinstance(next_char, bytes):
+                next_char = self.decoder.decode(next_char)
```
Not tested, but I'm not sure this will work for e.g. utf-8 data where decoding a single byte doesn't always make sense. As line 546 implies, this is only intended for things where `read()` will already return a `str` rather than `bytes` (apart from anything else, if bytes are being returned, the `next_char == ''` check on line 552 will never trigger and EOF won't end the attempts to read things...)

Might be better to check for a text encoding (or at simplest, do `assert isinstance(next_char, str)` after the `read()`)

codedstructure commented on pylibftdi/device.py:
```diff
@@ -543,24 +546,28 @@ def readline(self, size=0):
+        lines: list[str] = []
         if sizehint is not None:
             string_blob = self.read(sizehint)
+            if isinstance(string_blob, bytes):
+                string_blob = self.decoder.decode(string_blob)
```
See comments on lines 554-555 above - I think this is analogous and probably better to assume (e.g. with assert / type check and raise exception etc) that `read()` is returning `str`.

WhatTheFuzz replied:
Added the type check/assertion.

codedstructure reviewed (changes requested):
Thanks for these changes - overall I'm happy with the type hinting, but there are a few cases where functional changes (presumably intended to have consistent types when the original didn't) have broken things - though sadly not in a way the unit tests detect!

WhatTheFuzz committed:
Revert Driver to accept a list of strings.

- The Driver class should be able to load multiple libraries that
the user defines, not just one.
- Add tests for the Driver class. This currently does not work
with pytest because of the reassignment in test_common.py:
`pylibftdi.driver.Driver = MockDriver`. Seeking clarification on
how to best test the creation of a Driver object. Current testing
requires calling test_driver.py as a standalone program
(python3 ./test_driver.py).

WhatTheFuzz commented on pylibftdi/driver.py:
```diff
@@ -72,36 +73,43 @@ class Driver(object):
         """
         if libftdi_search is not None:
-            self._lib_search['libftdi'] = libftdi_search
+            self._lib_search['libftdi'] = list(libftdi_search)
```
Good catch, definitely not what I meant. I added a test case for this in `test_driver.py`. The initializer checks the type. If it's a string, create a list with a single element. If it's a list, replace the defaults with that list. If None, use the defaults. Raise a TypeError if it's none of those.

WhatTheFuzz committed:
Fix unit tests for Driver.

- Make _lib_search an instance variable. I can't see a specific
reason for it to be a class variable as only one should be created
at any point in time. Having one per instance has two benefits:
  - Makes unit tests easier.
  - Makes lookup times slightly faster because of Python scope
    searches checking local variables first (though likely
    inconsequential).
- Remove changing driver.Driver = MockDriver. The comment above
  says that it "does things", but I'm not sure what. The
  reassignment breaks unit tests, as MockDrivers are different
  than Drivers.

WhatTheFuzz committed:
Revert instance variable.

Driver is set up multiple times in the module and it may change
functionality in bigger ways than this PR is scoped for.

WhatTheFuzz commented on tests/test_common.py:
```diff
@@ -82,7 +82,6 @@ def libftdi_version(self):
 
 
 # importing this _does_ things...
-pylibftdi.driver.Driver = MockDriver
```
Could you clarify the purpose of the reassigned Driver class? I was able to omit it and still pass the unit tests, but I'm unsure if it affects anything else. Left in, the reassignment breaks the unit tests for `test_driver.py` since MockDriver has no attribute `_lib_search`.

WhatTheFuzz committed:
Fix broken Driver._load_library.

- Retype to Optional[list[str]].
- Add unittests for various _load_library scenarios. This may fail
  on certain systems if those libraries aren't present, though
  those cases should be rare.

WhatTheFuzz committed:
Require Python 3.7 or higher.

- Remove 3.6 from the classifiers.
- Add 3.11 to the classifiers.

WhatTheFuzz committed:
Add assertion that read() will return a string.

- .readlines() and .readline() should only be used for text mode,
  in which case .read() will return a string, not bytes.

WhatTheFuzz committed:
Add type annotations to Driver.

I missed this in testing before, but it won't work on 3.7 without
it. Unit tests passing 3.7.0 and 3.11.3.

codedstructure commented:
Thanks for this contribution to pylibftdi!
